numba/cuda/cudaimpl.py

```diff
@@ -405,6 +405,187 @@ def ptx_popc(context, builder, sig, args):
+    else:
+        raise errors.CudaLoweringError(f"Cuda target does not support "
+                                       f"from fp16 converting to float type "
+                                       f"with width {toty.bitwidth}-bits")
+
+    cvt2float_fnty = ir.FunctionType(context.get_value_type(toty),
+                                     [ir.IntType(16)])
+    cvt2float_asm = ir.InlineAsm(cvt2float_fnty,
+                                 f"cvt.{conv_type}.f16 $0, $1;",
+                                 f"={conv_constraint},h")
+    return builder.call(cvt2float_asm, [val])
+
+
+@lower_cast(types.IntegerLiteral, types.float16)
+def literal_int_to_number(context, builder, fromty, toty, val):
+    lit = context.get_constant_generic(
+        builder,
+        fromty.literal_type,
+        fromty.literal_value,
+    )
+    return context.cast(builder, lit, fromty.literal_type, toty)
+
+
+@lower_cast(types.Float, types.float16)
+def float_to_float16_cast(context, builder, fromty, toty, val):
+    if fromty.bitwidth == toty.bitwidth:
+        return val
+
+    if fromty.bitwidth == 32:
+        conv_type = "f32"
+        conv_constraint = "f"
+    elif fromty.bitwidth == 64:
+        conv_type = "f64"
+        conv_constraint = "d"
+    else:
+        raise errors.CudaLoweringError(f"Cuda target does not support "
+                                       f"converting from float type with width "
+                                       f" {toty.bitwidth}-bits to fp16")
+
+    cvt2float_fnty = ir.FunctionType(ir.IntType(16),
+                                     [context.get_value_type(fromty)])
+
+    cvt2float_asm = ir.InlineAsm(cvt2float_fnty,
+                                 f"cvt.rn.f16.{conv_type} $0, $1;",
+                                 f"=h,{conv_constraint}")
+    return builder.call(cvt2float_asm, [val])
+
+
+@lower_cast(types.float16, types.Integer)
+def float16_to_integer_cast(context, builder, fromty, toty, val):
+
+    if toty.bitwidth == 16:
+        conv_constraint = "h"
+    elif toty.bitwidth == 32:
+        conv_constraint = "r"
+    elif toty.bitwidth == 64:
+        conv_constraint = "l"
+    elif toty.bitwidth == 8:
+        conv_constraint = "c"
+    else:
+        raise errors.CudaLoweringError(f"Cuda target does not support "
+                                       f"from fp16 converting to integer type "
+                                       f"with width {toty.bitwidth}-bits")
+
+    if toty.signed:
+        signedness = "s"
+    else:
+        signedness = "u"
+
+    cvt2int_fnty = ir.FunctionType(context.get_value_type(toty),
+                                   [ir.IntType(16)])
+    cvt2int_asm = ir.InlineAsm(cvt2int_fnty,
+                               f"cvt.rni.{signedness}{toty.bitwidth}"
+                               f".f16 $0, $1;",
+                               f"={conv_constraint},h")
+    return builder.call(cvt2int_asm, [val])
+
+
+@lower_cast(types.Integer, types.float16)
```

Comment:
This can also lower the integer literal case:

```suggestion
@lower_cast(types.Integer, types.float16)
@lower_cast(types.IntegerLiteral, types.float16)
```

Reply:
Need to add a test that runs `cuda.fp16.hmul(x, 2)`. The `2` will need to be implicitly casted to `float16`.